### PR TITLE
2 enhancements to config.option

### LINKED ISSUE #51531: Docker login additional configuration options for docker authentication
https://github.com/saltstack/salt/blob/927adb31cbee132e9fba7fcd3e2ed7cd9643177c/salt/modules/dockermod.py#L1393

In reference to the above, would it be possible to make this function use master/minion config files _as well as_ pillar data?  I have used the "trifecta" of configuration options in other modules, but my external pillar setup does not easily support this particular configuration.  

I am running salt version 2018.3.2 on Centos 7.  My external pillar system is etcd.

### PR DESCRIPTION
**Pull request by terminalmage**
This makes the following two changes to `config.option`:

1. Matches are now returned from grains
2. Wildcard matches are now supported

For backward compatibility, wildcard matches are only attempted when `wildcard=True` is passed.

Additionally, the Docker registry configuration now uses `config.option`, meaning that registry auth can be configured elsewhere than the pillar data.

Resolves #51531.

### CODE REVIEW

**Comment by c4t3l** [comment]
+1

This will solve my original issue!  Thanks Erik.

**Comment by terminalmage** [comment]
It looks like Jenkins was in a bad state when I opened this, so since it's been a while I rebased and pushed to trigger new tests.

**Comment by terminalmage** [comment]
@Ch3LL I opened this 4 months ago and had added information to the Neon release notes, assuming that this would have been merged before we branched for Neon. Since that never happened, please let me know if you would like to either A) re-point this PR at Neon (will require a rebase from me, which I'm happy to do), or B) modify this PR to add the release notes to Sodium.

File: doc/topics/releases/neon.rst
Comment (waynew) [inline doc/topics/releases/neon.rst]: s/persistant/persistent

File: salt/modules/config.py
Comment (waynew) [inline salt/modules/config.py]: Would "reasonable defaults" read better here?

File: salt/modules/config.py
Comment (waynew) [inline salt/modules/config.py]: Would it be better to do

```
try:
    return __opts__[value]
except KeyError:
    pass
```
? (and then the same things for the later?)

Actually, if I understand this block correctly, couldn't you write it like this?

```
if not omit_opts:
    return __opts__.get(value, DEFAULTS.get(value, default))
```

Or perhaps

```
if not omit_opts:
    try:
        return __opts__[value]
    except KeyError:
        return DEFAULTS.get(value, default)
```

If you want to avoid always evaluating the `DEFAULTS.get(value, default)`.

**Review by waynew — CHANGES_REQUESTED** [review]

File: salt/modules/config.py
Comment (gtmanfred) [reply salt/modules/config.py]: exception handling is slower than just checking if the value is in the list.

and .get is also slower.

This is the fastest implementation

Comment (gtmanfred) [reply salt/modules/config.py]:
```
>>> import timeit
>>> setup = '__opts__ = {"one": 1, "two": 2}'
>>> timeit.timeit('''
... if 'one' in __opts__:
...     __opts__['one']
... ''', setup=setup)
0.047407952000000364
>>> timeit.timeit('''
... __opts__.get('one', 'THING')
... ''', setup=setup)
0.07143948800000999
```

Comment (waynew) [reply salt/modules/config.py]: Huh! I wonder if that's something that's changed or if I've just been misremembering that `get` was faster than `in`, and it's just better than a `try/except`.

Well, TIL!

👍

**Comment by Ch3LL** [comment]
ping @s0undt3ch or @twangboy any ideas why the windows tests are not finishing here?

**Comment by waynew** [comment]
@Ch3LL I think it might be the changes we got merged in yesterday. @terminalmage you mind rebasing on neon, I believe that will fix the issue 🤞

**Comment by terminalmage** [comment]
Given that the tests weren't even starting, and I didn't modify any files that affect Windows, I'm not sure I follow your logic. But I rebased again anyway, and sure enough, tests aren't even starting on Windows:

```
15:35:29 >>>>>>     Failed to complete #create action: [AMI 'ami-0b72a5e51f042d1bc' is pending, and cannot be run] on py2-windows-2016
```

By the way, you can see Jenkins' console output by clicking the "Console" link on the Jenkins test results.

**Comment by waynew** [comment]
I've had it happen where I'm trying to merge against say commit A on branch develop, and the build fails. Commits happen on develop and now we're at commit C. Rebasing (or merging C into my branch) triggers a rebuild, which will fix whatever. I'm not sure *why* GitHub says that a branch is out of date with the original... that seems pretty weird, now that you point it out :suspect: 

Regardless, it looks like now the only problem is this known Windows failure (Pedro is looking into it right now). Hopefully that shouldn't take *too* much.

**Comment by terminalmage** [comment]
> I've had it happen where I'm trying to merge against say commit A on branch develop, and the build fails. Commits happen on develop and now we're at commit C. Rebasing (or merging C into my branch) triggers a rebuild, which will fix whatever.

Again, the console was pretty clear that the failure was in just _starting_ the tests. There is actionable information that we can use, so I am honestly confused (and a little dismayed) that the plan seems to be to cross our fingers that "whatever" will be fixed by pushing and triggering another build.

> I'm not sure _why_ GitHub says that a branch is out of date with the original...

Really? You can see this change in the GitHub web interface when a pull request is merged into a branch. All the other PRs which were opened against that same branch are now out-of-date with the base branch since the base branch has commits that the head branch does not.

**Comment by waynew** [comment]
Ah, I had a fuzzy mental model going on, never mind 😝

> Again, the console was pretty clear that the failure was in just starting the tests. There is actionable information that we can use, so I am honestly confused (and a little dismayed) that the plan seems to be to cross our fingers that "whatever" will be fixed by pushing and triggering another build.

Apologies - I didn't actually check the console because I had been noticing failed windows builds for the last few days, and I knew that a couple of commits had gone in to fix other windows failures. I made an (incorrect) educated guess that this windows failure was related to the other failures. It wasn't until today that I learned of *this* particular class of failures. This was bad pattern matching on my part. Probably also driven by the fact that PRs have to be brought up-to-date before merging.

Next time I'll spend the time to ensure that the build failures are what I expect them to be.

**Comment by Ch3LL** [comment]
looks like the tests are passing. @waynew are we just waiting on your re-review?

**Review by waynew — APPROVED** [review]

**Pull request merged**
